## Constant and Quasi-constant features with Feature-engine

In this notebook, we will remove constant and quasi-constant features utilizing the new functionality from Feature-engine.

In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split

from feature_engine.selection import DropConstantFeatures

In [2]:
# load dataset

data = pd.read_csv('../dataset_1.csv')
data.shape

(50000, 301)

**Important**

In all feature selection procedures, it is good practice to select the features by examining only the training set. And this is to avoid overfit.

In [3]:
# separate dataset into train and test

X_train, X_test, y_train, y_test = train_test_split(
    data.drop(labels=['target'], axis=1), # drop the target
    data['target'], # just the target
    test_size=0.3,
    random_state=0)

X_train.shape, X_test.shape

((35000, 300), (15000, 300))

## Remove constant features

The DropConstantFeatures class from Feature-engine finds and removes constant and quasi-constant features from a dataset. We can remove constant features by setting the parameter tol to 1, or quasi-constant with smaller values for tol.

In [4]:
sel = DropConstantFeatures(tol=1, variables=None, missing_values='raise')

sel.fit(X_train)

DropConstantFeatures(missing_values='raise', tol=1,
                     variables=['var_1', 'var_2', 'var_3', 'var_4', 'var_5',
                                'var_6', 'var_7', 'var_8', 'var_9', 'var_10',
                                'var_11', 'var_12', 'var_13', 'var_14',
                                'var_15', 'var_16', 'var_17', 'var_18',
                                'var_19', 'var_20', 'var_21', 'var_22',
                                'var_23', 'var_24', 'var_25', 'var_26',
                                'var_27', 'var_28', 'var_29', 'var_30', ...])

In [5]:
# list of constant features

sel.features_to_drop_

['var_23',
 'var_33',
 'var_44',
 'var_61',
 'var_80',
 'var_81',
 'var_87',
 'var_89',
 'var_92',
 'var_97',
 'var_99',
 'var_112',
 'var_113',
 'var_120',
 'var_122',
 'var_127',
 'var_135',
 'var_158',
 'var_167',
 'var_170',
 'var_171',
 'var_178',
 'var_180',
 'var_182',
 'var_195',
 'var_196',
 'var_201',
 'var_212',
 'var_215',
 'var_225',
 'var_227',
 'var_248',
 'var_294',
 'var_297']

In [6]:
# number of constant features

len(sel.features_to_drop_)

34

In [7]:
# let's explore 1 of the constant feature values

X_train[sel.features_to_drop_[0]].unique()

array([0], dtype=int64)

In [8]:
# remove constant features from the data

X_train = sel.transform(X_train)
X_test = sel.transform(X_test)

X_train.shape, X_test.shape

((35000, 266), (15000, 266))

The datasets now contain 34 features less. 

## Remove quasi-constant features

In [9]:
sel = DropConstantFeatures(tol=0.998, variables=None, missing_values='raise')

sel.fit(X_train)

DropConstantFeatures(missing_values='raise', tol=0.998,
                     variables=['var_1', 'var_2', 'var_3', 'var_4', 'var_5',
                                'var_6', 'var_7', 'var_8', 'var_9', 'var_10',
                                'var_11', 'var_12', 'var_13', 'var_14',
                                'var_15', 'var_16', 'var_17', 'var_18',
                                'var_19', 'var_20', 'var_21', 'var_22',
                                'var_24', 'var_25', 'var_26', 'var_27',
                                'var_28', 'var_29', 'var_30', 'var_31', ...])

In [10]:
# number of quasi-constant features

len(sel.features_to_drop_)

108

In [11]:
# list of quasi-constant features

sel.features_to_drop_

['var_1',
 'var_2',
 'var_3',
 'var_6',
 'var_7',
 'var_9',
 'var_10',
 'var_11',
 'var_12',
 'var_14',
 'var_16',
 'var_20',
 'var_23',
 'var_24',
 'var_28',
 'var_32',
 'var_33',
 'var_34',
 'var_36',
 'var_39',
 'var_40',
 'var_42',
 'var_43',
 'var_44',
 'var_45',
 'var_48',
 'var_53',
 'var_56',
 'var_59',
 'var_60',
 'var_61',
 'var_65',
 'var_66',
 'var_67',
 'var_69',
 'var_71',
 'var_72',
 'var_73',
 'var_77',
 'var_78',
 'var_80',
 'var_81',
 'var_87',
 'var_89',
 'var_90',
 'var_92',
 'var_95',
 'var_97',
 'var_98',
 'var_99',
 'var_102',
 'var_104',
 'var_106',
 'var_111',
 'var_112',
 'var_113',
 'var_115',
 'var_116',
 'var_120',
 'var_122',
 'var_124',
 'var_125',
 'var_126',
 'var_127',
 'var_129',
 'var_130',
 'var_133',
 'var_135',
 'var_136',
 'var_138',
 'var_141',
 'var_142',
 'var_146',
 'var_149',
 'var_150',
 'var_151',
 'var_153',
 'var_158',
 'var_159',
 'var_167',
 'var_170',
 'var_171',
 'var_178',
 'var_180',
 'var_182',
 'var_183',
 'var_184',
 'var_187',


In [11]:
# percentage of observations showing each of the different values
# of the variable

var = sel.features_to_drop_[0]

X_train[var].value_counts() / np.float(len(X_train))

0    0.999629
3    0.000200
6    0.000143
9    0.000029
Name: var_1, dtype: float64

We can see that > 99% of the observations show one value, 0. Therefore, this features is fairly constant.

In [12]:
# let's explore another one

var = sel.features_to_drop_[2]

X_train[var].value_counts() / np.float(len(X_train))

0.0000         0.999629
35685.9459     0.000029
3583.3941      0.000029
15028.0560     0.000029
52105.7901     0.000029
10281.6000     0.000029
86718.0000     0.000029
207901.3365    0.000029
25905.4866     0.000029
5209.9500      0.000029
2641.0164      0.000029
12542.3100     0.000029
861.0900       0.000029
27.3000        0.000029
Name: var_3, dtype: float64

Go ahead and explore the rest of the quasi-constant variables.

We can then remove the quasi-constant features utilizing the transform() method. Feature-engine returns dataframes by default.

In [13]:
#remove the quasi-constant features

X_train = sel.transform(X_train)
X_test = sel.transform(X_test)

X_train.shape, X_test.shape

((35000, 158), (15000, 158))

By removing constant and almost constant features, we reduced the feature space from 300 to 158.

That is all for this lecture, I hope you enjoyed it and see you in the next one!